ddtrace/contrib/botocore/services/bedrock.py

```diff
@@ -292,61 +262,63 @@ def _extract_streamed_response_metadata(span: Span, streamed_body: List[Dict[str
-                prompt = v
-            if integration.is_pc_sampled_span(span):
-                v = integration.trunc(str(v))
-        span.set_tag_str("bedrock.request.{}".format(k), str(v))
-    return prompt
+        if k == "prompt" and ctx["bedrock_integration"].is_pc_sampled_llmobs(ctx[ctx["call_key"]]):
+            prompt = v
+    ctx.set_item("prompt", prompt)
 
 
 def handle_bedrock_response(
-    span: Span, integration: BedrockIntegration, result: Dict[str, Any], prompt: Optional[str] = None
+    ctx: core.ExecutionContext,
+    result: Dict[str, Any],
 ) -> Dict[str, Any]:
-    """Perform response param extraction and tagging."""
     metadata = result["ResponseMetadata"]
     http_headers = metadata["HTTPHeaders"]
-    span.set_tag_str("bedrock.response.id", str(metadata.get("RequestId", "")))
-    span.set_tag_str("bedrock.response.duration", str(http_headers.get("x-amzn-bedrock-invocation-latency", "")))
-    span.set_tag_str("bedrock.usage.prompt_tokens", str(http_headers.get("x-amzn-bedrock-input-token-count", "")))
-    span.set_tag_str("bedrock.usage.completion_tokens", str(http_headers.get("x-amzn-bedrock-output-token-count", "")))
 
-    # Wrap the StreamingResponse in a traced object so that we can tag response data as the user consumes it.
+    core.dispatch(
+        "botocore.patched_bedrock_api_call.success",
+        [
+            ctx,
+            str(metadata.get("RequestId", "")),
+            str(http_headers.get("x-amzn-bedrock-invocation-latency", "")),
+            str(http_headers.get("x-amzn-bedrock-input-token-count", "")),
+            str(http_headers.get("x-amzn-bedrock-output-token-count", "")),
+        ],
+    )
+
     body = result["body"]
-    result["body"] = TracedBotocoreStreamingBody(body, span, integration, prompt=prompt)
+    result["body"] = TracedBotocoreStreamingBody(body, ctx)
     return result
 
 
 def patched_bedrock_api_call(original_func, instance, args, kwargs, function_vars):
     params = function_vars.get("params")
-    trace_operation = function_vars.get("trace_operation")
-    operation = function_vars.get("operation")
     pin = function_vars.get("pin")
-    endpoint_name = function_vars.get("endpoint_name")
-    integration = function_vars.get("integration")
-    # This span will be finished once the user fully consumes the stream body, or on error.
-    bedrock_span = pin.tracer.trace(
-        trace_operation,
-        service=schematize_service_name("{}.{}".format(pin.service, endpoint_name)),
-        resource=operation,
+    model_provider, model_name = params.get("modelId").split(".")
+    with core.context_with_data(
+        "botocore.patched_bedrock_api_call",
+        pin=pin,
+        span_name=function_vars.get("trace_operation"),
+        service=schematize_service_name("{}.{}".format(pin.service, function_vars.get("endpoint_name"))),
+        resource=function_vars.get("operation"),
         span_type=SpanTypes.LLM,
```

Comment:
Just curious, how does this ensure the span.span_type gets set? I thought `context_with_data(..., **kwargs) only sets kwargs as part of the data dict

Reply:
@Yun-Kim The `span_type` gets pulled out of the context object [here](https://github.com/DataDog/dd-trace-py/blob/3ecf22943c4f9f0ab74b69242297b9ac3090129d/ddtrace/_trace/trace_handlers.py#L95) and is used as a kwarg to `start_span`[here](https://github.com/DataDog/dd-trace-py/blob/3ecf22943c4f9f0ab74b69242297b9ac3090129d/ddtrace/_trace/trace_handlers.py#L109). That `_start_span` handler is entered because there's a listener registered [here](https://github.com/DataDog/dd-trace-py/blob/3ecf22943c4f9f0ab74b69242297b9ac3090129d/ddtrace/_trace/trace_handlers.py#L674) for an event that gets fired on context creation [here](https://github.com/DataDog/dd-trace-py/blob/3ecf22943c4f9f0ab74b69242297b9ac3090129d/ddtrace/internal/core/__init__.py#L181).